### PR TITLE
Tooltip & Popover managers for event delegation

### PR DESCRIPTION
**Pull request by driskull**
## Add Tooltip & Popover managers for event delegation

Tooltip/Popover: create tool to initialize delegated events #397

If we can think of a better name for these two components that might be helpful. Their purpose is to setup delegated events for all tooltips/popovers contained within them.

## Changes

- Adds two new components to handle event delegation for Popovers and Tooltips.
  - A user just surrounds all reference elements they want to have events delegated with one of these components.
  - Previously, 4 events were being created per tooltip and one per popover. Now only 1 event per popover manager and 4 per tooltip manager.
  - This is much easier than having to import a utility and initialize it with an element.
  - `calcite-tooltip-manager` automatically adds hover/focus events for tooltips.
  - `calcite-popover-manager` automatically adds click event for popovers.
  - The selector `[data-calcite-tooltip-reference]` is used by default but can be customized by setting selector property.  Popover is `[data-calcite-popover-reference]`
- Fixed missed calcite background var name change in Popper styles
- Updated tests
- Updated stories

## Breaking Changes

- removes `add-click-handle` property from Popover. Users can now use `calcite-popover-manager` in its place.
- removes default event creation from tooltips. Users can now use `calcite-tooltip-manager` surrounding the referenced tooltip elements.

### CODE REVIEW

**Comment by driskull** [comment]
cc @AdelheidF @kevindoshier

**Comment by kevindoshier** [comment]
Sweet!

File: src/components/calcite-tooltip-manager/calcite-tooltip-manager.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  // --------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  // --------------------------------------------------------------------------
+
+  componentDidLoad() {
+    const { el } = this;
+
+    el.addEventListener("mouseenter", this.show, true);
```
Comment (jcfranco) [inline src/components/calcite-tooltip-manager/calcite-tooltip-manager.tsx]: Just an observation, but there's a lot in common with the popover manager here. Could the common stuff be extracted where the concrete managers would just need to register the events and handlers?

**Comment by driskull** [comment]
@jcfranco Great questions...

> Are managers expected to wrap the entire tooltip/popover content?

No, they just have to wrap the reference elements. This is to delegate the events for the reference elements only. so maybe the name of the component could be more obvious?

> Is the idea to have a single tooltip/popover manager in a consuming app? If not, what would happen if you have nested managers? For example, is the following valid: some-components use managers internally and some-app uses its own manager instance?

Nested managers are not recommended because then you would have multiple competing events.

However, you could have multiple managers that are not nested. Lets say you have a couple of sets of tooltips but they aren't ever displayed at the same time, you could use multiple managers in that case.

Its a balancing act really, not to have too many events but keep the scope of the delegated events as specific as possible.

@jcfranco does that make sense?

File: src/components/calcite-tooltip-manager/calcite-tooltip-manager.tsx (hunk 1 above)
Comment (driskull) [reply src/components/calcite-tooltip-manager/calcite-tooltip-manager.tsx]: yeah, I could share some code. 👍

**Comment by driskull** [comment]
> This may be tricky to enforce across different projects, so we'll have to keep an eye out.

Yep. Agreed. I think if they understand event delegation then it makes sense. We could have an explanation in the doc for how to use it.

Something like...

```
JavaScript event delegation is a simple technique by which you add a single event handler to a parent element in order to avoid having to add event handlers to multiple child elements.
```

Would the component name make more sense as something like:

- `calcite-popover-reference-manager`
- `calcite-popover-event-delegator`
- `calcite-popover-event-manager`

??

**Comment by driskull** [comment]
@macandcheese @paulcpederson can you take a look when you get a chance

**Comment by macandcheese** [comment]
To clarify - tooltips and popovers could continue to be used in their previous way - you don't necessarily need to wrap your content in `<...manager>` like in the dev demo examples?

**Comment by driskull** [comment]
> To clarify - tooltips and popovers could continue to be used in their previous way - you don't necessarily need to wrap your content in <...manager> like in the dev demo examples?

Correct. If you want to choose when these two open/close you can do that yourself by setting up your own events. However, I think most people would use these.

**Comment by driskull** [comment]
@macandcheese 

> tooltips and popovers could continue to be used in their previous way

Sort of. We were setting up the tooltip events for each tooltip previously so that is no longer the case. So it won't work as it was.

The popover had a property to setup the event. That was removed as well.

**Comment by driskull** [comment]
The components added in this PR are the equivalent of doing this in bootstrap:

```js
$(function () {
  $('[data-toggle="tooltip"]').tooltip()
})
```

tooltips and popovers could continue to be used in their previous way

**Comment by AdelheidF** [comment]
> Nested managers are not recommended because then you would have multiple competing events.

> However, I think most people would use these.

In the new map viewer the app is going to use these new managers. Does this mean that the components that are used inside the apps, such as PopupBuilder, smartMapping, FeatureFilter, ... should not use these new managers?

**Comment by driskull** [comment]
> In the new map viewer the app is going to use these new managers. Does this mean that the components that are used inside the apps, such as PopupBuilder, smartMapping, FeatureFilter, ... should not use these new managers?

No they should. 

We just want to try to avoid this:

```html
<calcite-tooltip-manager>
  <calcite-tooltip-manager></calcite-tooltip-manager>
</calcite-tooltip-manager>
```

There would be a couple ways to avoid this...

1. You only need the outer tooltip manager. The inner one is redundant.
2. You have different selectors for each tooltip manager so they target different elements to open/close.

**Pull request merged**